{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from graphviz import Digraph\n",
    "from matplotlib import pyplot as plt\n",
    "params, section = ['name', 'sequence', 'optional', 'quality'], 4\n",
    "fn = '6/s_6.first1000.fastq'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class Node():\n",
    "    def __init__(self, value=None):\n",
    "        self.value = value\n",
    "        \n",
    "    def get_value(self):\n",
    "        return self.value\n",
    "\n",
    "class DeBruijnGraph():\n",
    "    def __init__(self):\n",
    "        self.graph = {}\n",
    "        self.nodes = {}\n",
    "        self.edgeWeights = {}\n",
    "                \n",
    "    def add_vertex(self, node_str):\n",
    "        # this adds node as a vertex to graph\n",
    "        if node_str not in self.nodes:\n",
    "            node = Node(node_str)\n",
    "            self.nodes[node_str] = node\n",
    "            self.graph[node] = []\n",
    "            return node\n",
    "        else: return self.nodes[node_str]\n",
    "    \n",
    "    def _reverse(self, genome): #node1, node2):\n",
    "        complement = {'A':'T', 'C':'G', 'G':'C', 'T':'A'}\n",
    "        return ''.join(complement[i] for i in genome[::-1])\n",
    "    \n",
    "    def add_edge(self, node1, node2):\n",
    "        # adds an edge between node1 and node2 and their reverse\n",
    "        prefix_node = self.add_vertex(node1)\n",
    "        suffix_node = self.add_vertex(node2)\n",
    "        self.graph[prefix_node] += [suffix_node]\n",
    "        index = prefix_node.get_value() + suffix_node.get_value()[-1]\n",
    "        self.edgeWeights[index] = self.edgeWeights.get(index, 0) + 1\n",
    "        # add length to each time that this weight is updated.\n",
    "        \n",
    "    def first_node(self, node):\n",
    "        suffix_node = self.add_vertex(node)\n",
    "        count = [1 for n, value in self.graph.items() if suffix_node in value]\n",
    "        return sum(count) == 0\n",
    "    \n",
    "    def make_graph(self, genome, k=3):\n",
    "        for node1, node2 in kmer_generator(genome, k): # forward kmer\n",
    "            self.add_edge(node1, node2)\n",
    "#         for node1, node2 in kmer_generator(self._reverse(genome), k): # reverse kmer\n",
    "#             self.add_edge(node1, node2)\n",
    "\n",
    "class CondensedDeBruijnGraph(DeBruijnGraph):\n",
    "    def __init__(self):\n",
    "        self.graph = {} #from non-condensed\n",
    "        self.nodes = {} #from non-condensed\n",
    "        self.edgeWeights = {} #from non-condensed\n",
    "        self.condensedNodes = [] #{}\n",
    "        self.condensedGraph = {}\n",
    "    \n",
    "    def get_nodes(self):\n",
    "        return self.condensedNodes\n",
    "    \n",
    "    def get_edges(self):\n",
    "        for nodes, value in self.condensedGraph.items():\n",
    "            prefix_node, suffix_node, edge_label = nodes\n",
    "            yield (prefix_node.get_value(), suffix_node.get_value(), edge_label, value)\n",
    "    \n",
    "    def get_condensed_nodes(self):\n",
    "        for node in self.nodes: \n",
    "            prefix_node = self.add_vertex(node)\n",
    "            if len(self.graph[prefix_node])==0: # if this is the last element in the strand\n",
    "                self.condensedNodes.append(node)\n",
    "            elif len(self.graph[prefix_node])>1: # if this element has a branch, \\\n",
    "                #so number of edges leaving is greater than 1 (or entered from more than 1)\n",
    "                self.condensedNodes.append(node)\n",
    "            elif self.first_node(node): #this is the first element in the strand\n",
    "                self.condensedNodes.append(node)\n",
    "\n",
    "    def recurEdge(self, suffix_node, edge_label, weight, kmer_count):\n",
    "        edge_label += suffix_node.get_value()[-1]\n",
    "        if suffix_node.get_value() in self.condensedNodes:\n",
    "            if kmer_count <= 1.0:\n",
    "                return suffix_node, edge_label, 1.0, weight\n",
    "            return suffix_node, edge_label, weight/kmer_count, 1.0\n",
    "        else:\n",
    "            suffix_node_array = self.graph[suffix_node]\n",
    "            index = suffix_node.get_value() + suffix_node_array[0].get_value()[-1]\n",
    "            weight += self.edgeWeights[index]\n",
    "            return self.recurEdge(suffix_node_array[0], edge_label, weight, kmer_count+1.0)\n",
    "    \n",
    "    def make_condensed(self):\n",
    "        self.get_condensed_nodes()\n",
    "        for node in self.condensedNodes: \n",
    "            prefix_node = self.add_vertex(node)\n",
    "            suffix_node_array = self.graph[prefix_node]     \n",
    "            for suffix_node in suffix_node_array:\n",
    "                index = node + suffix_node.get_value()[-1]\n",
    "                weight = self.edgeWeights[index]\n",
    "                cond_node, edge_label, avg_cov, length = self.recurEdge(suffix_node, node, \\\n",
    "                                                                               weight, 1.0)\n",
    "                self.condensedGraph[(prefix_node, cond_node, edge_label)] = avg_cov, length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_reads(fn):\n",
    "    with open(fn, 'r') as file:\n",
    "        lines = []\n",
    "        for line in file:\n",
    "            lines.append(line.rstrip())\n",
    "            if len(lines) == section:\n",
    "                record = {k: v for k, v in zip(params, lines)}\n",
    "                lines = []\n",
    "                yield record['sequence']\n",
    "\n",
    "# write k-mer of graph\n",
    "def kmer_generator(genome, k=3):\n",
    "    i = 0\n",
    "    while i + k < len(genome):\n",
    "        node1, node2 = genome[i:i+k], genome[i+1:i+1+k]\n",
    "        i += 1\n",
    "        yield node1, node2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Condensed De Brujin Analysis\n",
    "k = 55\n",
    "genomes = ['AAACTTATGGGACCCA', 'AAACGGGATTATCCCA', 'AAGACTCCGACTGGGACTTT', 'ACGTCCGTAA']\n",
    "graph = CondensedDeBruijnGraph()\n",
    "for genome in get_reads(fn): #get_reads(fn): #genomes[2:]: #\n",
    "    graph.make_graph(genome, k)\n",
    "graph.make_condensed()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'graph_55mer.pdf'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_edges = [(u, v, w, x[0], x[1]) for u, v, w, x in graph.get_edges()]\n",
    "all_edges = list(set(all_edges))\n",
    "all_nodes = graph.get_nodes()\n",
    "dot = Digraph() #Diagraph(format='png')\n",
    "dict_keys = {all_nodes[i]:str(i) for i in range(len(all_nodes)) } \n",
    "[dot.node(dict_keys[i], i) for i in all_nodes]\n",
    "[dot.edge(dict_keys[i], dict_keys[v], label=w+' ('+'{:.0f}'.format(length)+', '+'{:.2f}'.format(cov)+')') \\\n",
    "                                             for i, v, w, cov, length in all_edges]\n",
    "# [dot.edge(dict_keys[i], dict_keys[v]) for i, v, _, _, _ in all_edges]\n",
    "dot.render('graph_'+str(k)+'mer', view=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Graph simplification: Introducing tip removal - Remove short length and small coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "length = [y for _,_,_,x,y in all_edges]\n",
    "coverage = [x for _,_,_,x,y in all_edges]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAAEYCAYAAAAJeGK1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4yLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvNQv5yAAAE2FJREFUeJzt3X+MXedd5/H3BycpP+tCHVBlO2uD\nrcCobJvdIcmqi8p2WzRulDqgqrWFWlSsWNklEUj8sWZVUaHtHwl/wMqsaeRuLBdUxXVDGxziKK0g\nVQRNS9wAxa43MM22ykTdtSGtoaDFuP3uH3O8ezvMeGbsmXufc+/7JVmZ89xzz/0+uXP98XOe55yb\nqkKSpNZ826gLkCRpMQaUJKlJBpQkqUkGlCSpSQaUJKlJBpQkqUkGlCSpSQaUJKlJ1426gCtJchdw\nB/BK4KGq+sSIS5IkDcmyI6gk357kT5L8eZIzSX71al8syZEk55KcXuSxmSTPJ5lNcgCgqh6tqruB\ne4B3Xu3rSpL6ZyWn+P4ReFNVvQ54PTCT5PbBHZJ8f5LvWdC2Y5FjHQVmFjYm2QAcAnYBU8DeJFMD\nu7y3e1ySNCGWDaia9/Vu8/ruz8Ib+L0ReDTJKwCS3A385iLHehp4eZGXuRWYraoXquoicAzYnXkP\nAE9U1XOL1ZfkziSHk9y5XF8kSf2xojmoboTzOWAHcKiqPjv4eFV9NMl24CNJPgr8HPCWVdSxGXhx\nYHsOuA24D3gzsDHJjqp6cOETq+ox4LFNmzbV9PT0Kl5SWjuf+9zn/rqqbhx1HSu1adOm2rZt26jL\n0IRa6edlRQFVVd8AXp/kVcDHk7y2qk4v2OfXkhwDPgD80MCo66pV1UHg4Er23bZtG6dOnbrWl5Su\nSpIvj7qG1fDzolFa6edlVcvMq+prwFMsPo/048BrgY8D71vNcYGXgK0D21u6NknShFrJKr4bu5ET\nSb6D+VN3/2PBPrcAh4HdwHuAVyd5/yrqeBbYmWR7khuAPcCJVTxfkjRmVjKCeg3wVJLPMx8kn6yq\n31+wz3cC76iqL1bVN4F3A/9sCJfkYeAZ4OYkc0n2AVTVJeBe4EngLHC8qs5cbackSf237BxUVX0e\nuGWZff54wfY/AR9cZL+9VzjGSeDkcvVIkiaDtzqSJDXJgJIkNcmAkiQ1yYCSJsjlO69cuHBh1KVI\nyzKgpAlSVY9V1f6NGzeOuhRpWU1/3cZa2Xbg8at+7pfuv2MNK5Ha5+dFrXAEJUlqkgElSWqSASVJ\napIBJUlq0kQsklCbrnYy3ol4aTI4gpIkNcmAkiQ1yYCSJDXJgJIkNcmAkiQ1yYCSJDXJgJIkNcnr\noAR4TZKk9jiCkiQ1yYCSei7JjyR5MMkjSf7DqOuR1ooBJTUoyZEk55KcXtA+k+T5JLNJDgBU1dmq\nugd4B/CGUdQrrQcDSmrTUWBmsCHJBuAQsAuYAvYmmeoeexvwOHByuGVK68eAkhpUVU8DLy9ovhWY\nraoXquoicAzY3e1/oqp2AT+z1DGT7E9yKsmp8+fPr1fp0ppxFZ/UH5uBFwe254DbkvwE8NPAK7jC\nCKqqDgOHAaanp2v9ypTWhgEl9VxVfQr41IjLkNacp/ik/ngJ2DqwvaVrW7EkdyY5fOHChTUtTFoP\nBpTUH88CO5NsT3IDsAc4sZoDVNVjVbV/48aN61KgtJYMKKlBSR4GngFuTjKXZF9VXQLuBZ4EzgLH\nq+rMKOuU1pNzUFKDqmrvEu0ncSm5JoQjKGmCOAelPjGgpAniHJT6xICSJDXJgJIkNcmAkiaIc1Dq\nEwNKmiDOQalPDChJUpMMKElSkwwoaYI4B6U+MaCkCeIclPrEgJIkNcmAkiQ1yYCSJDXJgJImiIsk\n1CcGlDRBXCShPjGgJElNMqAkSU0yoCRJTTKgJElNMqAkSU0yoKQJ4jJz9YkBJU0Ql5mrTwwoSVKT\nrht1AVo72w48PuoSJGnNOIKSJDXJgJIkNcmAkiQ1yYCSJDXJgJImiNdBqU8MKGmCeB2U+sRl5rom\nLm2XtF4cQUmSmmRASZKaZEBJkppkQEmSmtT0IokkdwF3AK8EHqqqT4y4JEnSkCw7gkqyNclTSb6Q\n5EySX7jaF0tyJMm5JKcXeWwmyfNJZpMcAKiqR6vqbuAe4J1X+7qSpP5ZySm+S8AvVdUUcDvw80mm\nBndI8v1JvmdB245FjnUUmFnYmGQDcAjYBUwBexe8xnu7xyVJE2LZgKqqr1TVc93PfwecBTYv2O2N\nwKNJXgGQ5G7gNxc51tPAy4u8zK3AbFW9UFUXgWPA7sx7AHjicg2SpMmwqjmoJNuAW4DPDrZX1UeT\nbAc+kuSjwM8Bb1nFoTcDLw5szwG3AfcBbwY2JtlRVQ8uUtOdwJ07diw2YJMk9dWKV/El+W7gd4Ff\nrKq/Xfh4Vf0a8H+ADwBvq6qvX2txVXWwqv51Vd2zWDh1+3jrFmmFvBef+mRFAZXkeubD6cNV9bEl\n9vlx4LXAx4H3rbKOl4CtA9tbujZJa8h/0KlPVrKKL8BDwNmq+vUl9rkFOAzsBt4DvDrJ+1dRx7PA\nziTbk9wA7AFOrOL5kqQxs5IR1BuAdwFvSvJn3Z+3LtjnO4F3VNUXq+qbwLuBLy88UJKHgWeAm5PM\nJdkHUFWXgHuBJ5lfhHG8qs5cda8kSb237CKJqvojIMvs88cLtv8J+OAi++29wjFOAieXq0eSNBm8\n1ZEkqUkGlCSpSQaUJKlJBpQkqUkGlCSpSQaUJKlJBpQkqUkGlCSpSQaUJKlJBpQkqUkGlCSpSQaU\nJKlJq/pGXUltSnIXcAfwSuChqvrEiEuSrpkjKKlRSY4kOZfk9IL2mSTPJ5lNcgCgqh6tqruBe4B3\njqJeaa0ZUFK7jgIzgw1JNgCHgF3AFLA3ydTALu/tHpd6z4CSGlVVTwMvL2i+FZitqheq6iJwDNid\neQ8AT1TVc8OuVVoPBpTUL5uBFwe257q2+4A3A29Pcs9iT0yyP8mpJKfOnz+//pVK18hFEtIYqKqD\nwMFl9jkMHAaYnp6uYdQlXQtHUFK/vARsHdje0rVJY8eAkvrlWWBnku1JbgD2ACdW+uQkdyY5fOHC\nhXUrUForBpTUqCQPA88ANyeZS7Kvqi4B9wJPAmeB41V1ZqXHrKrHqmr/xo0b16doaQ05ByU1qqr2\nLtF+Ejg55HKkoXMEJU0QT/GpTwwoaYJ4ik99YkBJkppkQEmSmmRASRPEOSj1iQElTRDnoNQnBpQk\nqUkGlCSpSQaUJKlJBpQ0QVwkoT4xoKQJ4iIJ9YkBJUlqkgElSWqSASVJapIBJU0QF0moTwwoaYK4\nSEJ9YkBJkppkQEmSmmRASZKaZEBJkppkQEmSmmRASRPEZebqEwNKmiAuM1efGFCSpCYZUJKkJl03\n6gLG1bYDj1/1c790/x1rWIkk9ZMjKElSkwwoSVKTDChJUpMMKGmCeB2U+sSAkiaI10GpTwwoSVKT\nDChJUpO8DqpB13IN1STwGjNpMjiCkiQ1yYCSJDXJgJIkNcmAkiQ1yYCSJDXJgJIkNcmAkiQ1yYCS\nJoj34lOfGFDSBPFefOoTA0qS1CQDSpLUJANKktQkA0qS1CQDSpLUJANKktQkA0qS1CQDSpLUJANK\nktQkA0qS1CQDSpLUJANKktQkA0qS1CQDSpLUJANKktQkA0rquSQ/mOShJI+MuhZpLRlQUoOSHEly\nLsnpBe0zSZ5PMpvkAEBVvVBV+0ZTqbR+DCipTUeBmcGGJBuAQ8AuYArYm2Rq+KVJw2FASQ2qqqeB\nlxc03wrMdiOmi8AxYPdKj5lkf5JTSU6dP39+DauV1ocBJfXHZuDFge05YHOSVyd5ELglyS8v9eSq\nOlxV01U1feONN653rdI1u27UBUi6NlX1N8A9o65DWmuOoKT+eAnYOrC9pWtbsSR3Jjl84cKFNS1M\nWg8GlNQfzwI7k2xPcgOwBzixmgNU1WNVtX/jxo3rUqC0lgwoqUFJHgaeAW5OMpdkX1VdAu4FngTO\nAser6swo65TWk3NQUoOqau8S7SeBk0MuRxoJR1DSBHEOSn1iQEkTxDko9YkBJUlqkgElSWqSASVN\nEOeg1CcGlDRBnINSnxhQkqQmGVCSpCYZUNIEcQ5KfWJASRPEOSj1iQElSWqSASVJapIBJUlqkgEl\nSWqSASVNEFfxqU8MKGmCuIpPfWJASZKaZEBJkppkQEmSmmRASZKaZEBJE8RVfOoTA0qaIK7iU58Y\nUJKkJhlQkqQmGVCSpCYZUJKkJhlQkqQmGVCSpCYZUJKkJjUZUEnuSvLBJB9J8pOjrkcaF16oqz4Z\nWkAlOZLkXJLTC9pnkjyfZDbJAYCqerSq7gbuAd45rBqlceeFuuqTYY6gjgIzgw1JNgCHgF3AFLA3\nydTALu/tHpckTZihBVRVPQ28vKD5VmC2ql6oqovAMWB35j0APFFVzw2rRklSO64b8etvBl4c2J4D\nbgPuA94MbEyyo6oeXOzJSfYD+wFuuummdS5V42Dbgcev6nlfuv+ONa5E0nJGHVCLqqqDwMEV7HcY\nOAwwPT1d612XJGl4Rr2K7yVg68D2lq5NkjThRh1QzwI7k2xPcgOwBzgx4pokSQ0Y5jLzh4FngJuT\nzCXZV1WXgHuBJ4GzwPGqOjOsmiRJ7RraHFRV7V2i/SRwclh1SJL6YdSn+CRJWpQBJU0Qb3WkPjGg\npAnirY7UJwaUJKlJBpQkqUkGlCSpSQaUJKlJBpQkqUkGlCSpSQaUJKlJBpQkqUkGlCSpSQaUJKlJ\nvQ8o7y0mSeOp9wHlvcUkaTz1PqAkSePJgJIkNcmAkiQ1aWhf+S5pfST5LuC3gIvAp6rqwyMuSVoT\njqCkBiU5kuRcktML2meSPJ9kNsmBrvmngUeq6m7gbUMvVlonBpTUpqPAzGBDkg3AIWAXMAXsTTIF\nbAFe7Hb7xhBrlNaVASU1qKqeBl5e0HwrMFtVL1TVReAYsBuYYz6kwM+0xohzUMvYduDxUZcgXbaZ\n/z9Sgvlgug04CPy3JHcAjy315CT7gf0AN9100zqWqSu52r9TvnT/HWtcyfKu5e+/tajXgJJ6rqr+\nHnjPCvY7DBwGmJ6ervWuS7pWng6Q+uMlYOvA9pauTRpLBpTUH88CO5NsT3IDsAc4sZoDeO9K9YkB\nJTUoycPAM8DNSeaS7KuqS8C9wJPAWeB4VZ1ZzXG9d6X6xDkoqUFVtXeJ9pPAySGXI42EIyhpgniK\nT31iQEkTxFN86hMDSpLUJANKktQkA0qaIM5BqU9SNR4XlCc5D3x5iYc3AX89xHLWy7j0A8anL5f7\n8S+q6sZRF7NSE/J5WYlJ6Wtr/VzR52VsAupKkpyqqulR13GtxqUfMD59GZd+DBrHPi1lUvra1356\nik+S1CQDSpLUpEkJqMOjLmCNjEs/YHz6Mi79GDSOfVrKpPS1l/2ciDkoSVL/TMoISpLUMwaUJKlJ\nYx1QSWaSPJ9kNsmBUdeznCRHkpxLcnqg7fuSfDLJX3X//d6uPUkOdn37fJJ/NbrKv1WSrUmeSvKF\nJGeS/ELX3qu+JPn2JH+S5M+7fvxq1749yWe7ej/SfTcTSV7Rbc92j28bZf3LWez3bcHjTb4vq7WC\nfv5M17+/SPLpJK8bdo1rYbl+Duz3Y0kuJXn7sGq7WmMbUEk2AIeAXcAUsDfJ1GirWtZRYGZB2wHg\nD6pqJ/AH3TbM92tn92c/8IEh1bgSl4Bfqqop4Hbg57v/933ryz8Cb6qq1wGvB2aS3A48APxGVe0A\nvgrs6/bfB3y1a/+Nbr+WHeWf/74NavV9Wa2jXLmf/xN4Y1X9KPBf6OmCApbv5+W/Fx8APjGMgq7V\n2AYUcCswW1UvVNVF4Biwe8Q1XVFVPQ28vKB5N/Ch7ucPAXcNtP92zfsM8KokrxlOpVdWVV+pque6\nn/+O+S/X20zP+tLV8/Vu8/ruTwFvAh7p2hf243L/HgH+fZIMqdxVW+L3bVCT78tqLdfPqvp0VX21\n2/wMsGUoha2xFbyfAPcBvwucW/+Krt04B9Rm4MWB7bmurW9+oKq+0v38v4Af6H7uRf+601y3AJ+l\nh31JsiHJnzH/gf4k8EXga92328K31vr/+tE9fgF49XArXlPNvi/raB/wxKiLWA9JNgM/RY9Gwn6j\nbo9UVSXpzXUBSb6b+X+t/WJV/e3gYKIvfamqbwCvT/Iq4OPAD4+4JK2TJP+O+YD6t6OuZZ38V+A/\nVdU3Gx7Yf4txDqiXgK0D21u6tr7530leU1Vf6U6vXB6aN92/JNczH04frqqPdc297AtAVX0tyVPA\nv2H+VNd13ShpsNbL/ZhLch2wEfibkRS8Npp/X9ZKkn8J/HdgV1X1+T27kmngWBdOm4C3JrlUVY+O\ntqyljfMpvmeBnd2KqxuAPcCJEdd0NU4AP9v9/LPA7w20v7tbaXU7cGHg9NlIdfMuDwFnq+rXBx7q\nVV+S3NiNnEjyHcBbmJ9Pewq4vAJqYT8u9+/twB9Wv6+Eb/J9WWtJbgI+Bryrqv5y1PWsl6raXlXb\nqmob83Ok/7HlcIIxHkFV1aUk9wJPAhuAI1V1ZsRlXVGSh4GfADYlmQPeB9wPHE+yj/mvR3hHt/tJ\n4K3ALPAPwHuGXvDS3gC8C/iLbv4G4D/Tv768BvhQt/Lp24DjVfX7Sb7A/L9E3w/8KfNhTPff30ky\ny/xk9Z5RFL1SS/y+XQ9QVQ/S7vuyKivo568wP1f4W93o4lIf7/y9gn72jrc6kiQ1aZxP8UmSesyA\nkiQ1yYCSJDXJgJIkNcmAkiQ1yYCSJDXJgJIkNen/Ao89RISVXa8TAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x109e1a0b8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.subplot(121)\n",
    "plt.hist(length, log=True)\n",
    "plt.subplot(122)\n",
    "plt.hist(coverage, log=True)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### I removed all the tips. Noticed that tips overwhelmingly have a length of 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'graph_55mer_notips.pdf'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_edges = [(u, v, w, cov, length) for u, v, w, cov, length in all_edges if length > 1]\n",
    "dot = Digraph() #Diagraph(format='png')\n",
    "[dot.edge(dict_keys[i], dict_keys[v], label=w+' ('+'{:.0f}'.format(length)+', '+'{:.2f}'.format(cov)+')') \\\n",
    "                                             for i, v, w, cov, length in no_edges]\n",
    "# [dot.edge(dict_keys[i], dict_keys[v]) for i, v, _, _, _ in all_edges]\n",
    "dot.render('graph_'+str(k)+'mer_notips', view=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Graph simplification: Introducing tip removal - Remove low coverage and short edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'graph_55mer_lowCov.pdf'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_edges = [(u, v, w, cov, length) for u, v, w, cov, length in all_edges if length > 250]\n",
    "dot = Digraph()\n",
    "[dot.edge(dict_keys[i], dict_keys[v], label=w+' ('+'{:.0f}'.format(length)+', '+'{:.2f}'.format(cov)+')') \\\n",
    "                                             for i, v, w, cov, length in no_edges]\n",
    "# [dot.edge(dict_keys[i], dict_keys[v]) for i, v, _, _, _ in all_edges]\n",
    "dot.render('graph_55'+str(k)+'mer_lowCov', view=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Removing low coverage and short edges results in a disconnected graph so I attempted to recreate a condensed graph after removal. This resulted in a more streamlined graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Try a shorter sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Condensed De Brujin Analysis\n",
    "k = 21\n",
    "graph = CondensedDeBruijnGraph()\n",
    "for genome in get_reads(fn): #get_reads(fn): #genomes[2:]: #\n",
    "    graph.make_graph(genome, k)\n",
    "graph.make_condensed()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'graph_forward21mer.png'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_edges = [(u, v, w, x[0], x[1]) for u, v, w, x in graph.get_edges()]\n",
    "all_edges = list(set(all_edges))\n",
    "all_nodes = graph.get_nodes()\n",
    "dot = Digraph(format='png') #Diagraph(format='png')\n",
    "dict_keys = {all_nodes[i]:str(i) for i in range(len(all_nodes)) } \n",
    "# [dot.node(dict_keys[i], i) for i in all_nodes]\n",
    "[dot.edge(dict_keys[i], dict_keys[v], label=w+' ('+'{:.0f}'.format(length)+', '+'{:.2f}'.format(cov)+')') \\\n",
    "                                             for i, v, w, cov, length in all_edges]\n",
    "# [dot.edge(dict_keys[i], dict_keys[v]) for i, v, _, _, _ in all_edges]\n",
    "dot.render('graph_forward'+str(k)+'mer', view=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
